includes/admin/extensions/abstract-extension.php

```diff
@@ -0,0 +1,269 @@
+				$button = array(
+					/* translators: The extension name. */
+					'button_text' => sprintf( __( 'Log In to Your Account to Download %s', 'easy-digital-downloads' ), $product_data['title'] ),
+					'href'        => $this->get_upgrade_url( $product_data, $item_id, true ),
+					'new_tab'     => true,
+					'type'        => $type,
+				);
+			} else {
+				$button = array(
+					/* translators: The extension name. */
+					'button_text' => sprintf( __( 'Upgrade Today to Access %s!', 'easy-digital-downloads' ), $product_data['title'] ),
+					'href'        => $this->get_upgrade_url( $product_data, $item_id ),
+					'new_tab'     => true,
+					'type'        => $type,
+				);
+			}
+		} elseif ( ! $this->manager->is_plugin_active( $product_data['basename'] ) ) {
+			// If the extension is installed, but not activated, the button will prompt to activate it.
+			$button['plugin'] = $product_data['basename'];
+			$button['action'] = 'activate';
+			/* translators: The extension name. */
+			$button['button_text'] = sprintf( __( 'Activate %s', 'easy-digital-downloads' ), $product_data['title'] );
+		}
+
+		return $button;
+	}
+
+	/**
+	 * Gets the upgrade URL for the button.
+	 *
+	 * @todo add UTM parameters
+	 * @since 2.11.x
+	 * @param array $product_data The array of provided data about the extension.
+	 * @param int   $item_id      The item/product ID.
+	 * @param bool  $has_access   Whether the user already has access to the extension (based on pass level).
+	 * @return string
+	 */
+	private function get_upgrade_url( $product_data, $item_id, $has_access = false ) {
+		$url            = 'https://easydigitaldownloads.com';
+		$tab            = ! empty( $_GET['tab'] ) ? sanitize_text_field( $_GET['tab'] ) : '';
+		$utm_parameters = array(
+			'p'            => urlencode( $item_id ),
+			'utm_source'   => 'settings',
+			'utm_medium'   => urlencode( $tab ),
+			'utm_campaign' => 'admin',
+			'utm_term'     => urlencode( $product_data['slug'] ),
+		);
+
+		if ( $has_access ) {
+			$url = 'https://easydigitaldownloads.com/your-account/your-downloads/';
+			unset( $utm_parameters['p'] );
+		} elseif ( ! empty( $product_data['upgrade_url'] ) ) {
+			$url = esc_url( $product_data['upgrade_url'] );
+			unset( $utm_parameters['p'] );
+		}
+
+		return add_query_arg(
+			$utm_parameters,
+			$url
+		);
+	}
+
+	/**
+	 * Gets the array of parameters for the link to configure the extension.
+	 *
+	 * @since 2.11.x
+	 * @param array  $product_data  The array of provided data about the extension.
+	 * @return array
+	 */
+	protected function get_link_parameters( $product_data ) {
+		return array(
+			/* translators: The extension name. */
+			'button_text' => sprintf( __( 'Configure %s', 'easy-digital-downloads' ), $product_data['title'] ),
+			'href'        => add_query_arg(
+				array(
+					'post_type' => 'download',
+					'page'      => 'edd-settings',
+					'tab'       => $product_data['tab'],
+					'section'   => $product_data['section'],
```

Comment:
```suggestion
					'tab'       => urlencode( $product_data['tab'] ),
					'section'   => urlencode( $product_data['section'] ),
```

Reply:
I will commit this but I want to leave it as a placeholder for half a moment. I need to think about how to handle extensions which may not have settings at all...I don't know if we have any but if we do, I think this needs to be handled differently.